### ISSUE TITLE
Return value of jwt_verify() is an accident waiting to happen

### ISSUE THREAD

**Issue by TimWolla** [haproxy/haproxy#1419]
### Detailed Description of the Problem

The `jwt_verify` converter returns 0 for invalid JWTs, 1 for valid and > 1 for other errors. If a user does not properly read the documentation, then they might do the following:

```
http-request deny unless { var(txn.bearer),jwt_verify(txn.jwt_alg,test) -m bool }
```

instead of

```
http-request deny unless { var(txn.bearer),jwt_verify(txn.jwt_alg,test) 1 }
```

Using the `bool` matching method is not completely unreasonable given the name of the converter. However this will accept invalid JWTs!

### Expected Behavior

It should be hard to misuse this (security critical) converter.

### Steps to Reproduce the Behavior

1. Look at the documentation.
2. Send a token with a non-b64 signature, such as `eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.5mhBHqs5_DTLdINd9p5m7ZJ6XD0Xc55kIaCRY5r6HR?` (with the literal question mark).

### Do you have an idea how to solve the issue?

The valid state should be redefined to `0`, because then using the `-m bool` fetch very obviously doesn't work while conducting functional tests, forcing the user to look into the docs.

### What is your configuration?

```haproxy
listen foo
	mode http
	bind *:8080

	http-request set-var(txn.bearer) http_auth_bearer
	http-request set-var(txn.jwt_alg) var(txn.bearer),jwt_header_query('$.alg')
	http-request deny unless { var(txn.jwt_alg) "HS256" }
	http-request deny unless { var(txn.bearer),jwt_verify(txn.jwt_alg,test) -m found }

	http-request return status 200
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.5-dev10 2021/10/16 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Running on: Linux 5.4.0-89-generic #100-Ubuntu SMP Fri Sep 24 14:50:10 UTC 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_OPENSSL=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL +THREAD +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL -LUA +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC -PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
Running on OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Support for malloc_trim() is enabled.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built without PCRE or PCRE2 support (using libc's regex instead)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.3.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : none

Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace
```

**Comment by wtarreau** [comment]
While I'm also a bit worried with the confusing ordering of (bad,good,bad), I don't think there is any way to solve the "-m bool" one because in your example, someone using -m bool would have it reversed, naturally believing that "jwt_verify" returns true when the token is verified while it's the opposite, thus the "-m bool" is doomed in all situations. However what could at least be done would be to have <0 for errors, 0 for not correct, >0 for OK. Or we could find a different name that doesn't imply a possible boolean or one that means the opposite (e.g. jwt_decode_error(), though I find this awful). At least with <=0 vs >0 we can provide examples centered around zero and put a big fat warning about the risks that a cast to bool represents.

**Comment by TimWolla** [comment]
> I don't think there is any way to solve the "-m bool" one because in your example, someone using -m bool would have it reversed, naturally believing that "jwt_verify" returns true when the token is verified while it's the opposite, thus the "-m bool" is doomed in all situations.

Yes, I understand that `-m bool` would be impossible to use. However it would be very obvious that it does not work: If the admin tests with a valid token and it doesn't work then they'll probably look into the docs.

**Comment by rlebreton** [comment]
Not so sure we could qualify as a bug such a case in which a user would choose to use a security critical converter without reading its documentation, which explicitly tells in the first paragraph that it returns 1 in case of verification success and has a quite visible array in its description in which all the possible error codes are detailed.
It still submitted a patch to @wlallemand to switch all the error codes to negative values and I mentioned in the documentation that the return value should not be converted to boolean. None of that would prevent someone from misusing the converter though.

**Comment by wtarreau** [comment]
Thanks Rémi. We can never prevent anyone from shooting themselves in the foot, but it's true that as long as we can help avoid bugs in configs it's better. I'm not so much worried about those creating a config from scratch than those troubleshooting configs in field by disabling tests or inverting them (or copying examples found on the net, very common). If at least examples start to appear with "le 0" or "lt 0" or "gt 0" that will catch the attention and avoid most common mistakes I guess.

**Comment by wlallemand** [comment]
Fixed in 1c891bcc90346aa84f8b16bc6fe00fa6c104a4cb.
